### PR TITLE
Cache Serialize API

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Removes the ExtendReservation API from the datacatalog protobuf definition.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
The datacatalog ExtendReservation API is continually called when an Artifact is already reserved and that reservation needs to be extended past the current expiration. We are changing the implementation of the GetOrReserveArtifact API to extend a reservation if valid. Therefore, ExtendReservation is no longer necessary. We have included additional fields into the GetOrReserveArtifactResponse to reflect these changes.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/267
https://github.com/flyteorg/flyte/issues/872

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/core/catalog.proto [hunk 1]
```diff
@@ -41,3 +41,17 @@ message CatalogMetadata {
        TaskExecutionIdentifier source_task_execution = 3;
     }
 };
+
+// Indicates the status of a catalog reservation operation.
+enum CatalogReservationStatus {
```
Comment (EngHabu) [inline protos/flyteidl/core/catalog.proto:46]: You may want to move this into a message... because of the way code generation happens, this will end up with weird naming in some languages...

**Review by EngHabu — DISMISSED** [review]
Finally started looking at the cache serialization PRs... 
Let's wait on merging them until they are all +1'd

File: protos/flyteidl/core/catalog.proto (hunk 1 above)
Comment (hamersaw) [reply protos/flyteidl/core/catalog.proto:46]: @EngHabu  We'll want to move the CatalogCacheStatus enum at the top of this file then as well. Do we just want to move all of this file out of the core package and into the datacatalog package?

Comment (EngHabu) [reply protos/flyteidl/core/catalog.proto:46]: Example: https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/core/execution.proto

I don't think you need to do the refactoring right now... we need a bigger project to revisit DataCatalog design and we can do the cleanup as part of that... but at least for new additions (like this PR), let's prevent future cleanup

Comment (hamersaw) [reply protos/flyteidl/core/catalog.proto]: sounds great, done.

**Review by EngHabu — DISMISSED** [review]

File: protos/flyteidl/core/tasks.proto [hunk 2]
```diff
@@ -92,6 +92,9 @@ message TaskMetadata {
     oneof interruptible_value {
         bool interruptible = 8;
     };
+
+    // Indicates whether the system should attempt to execute discoverable instances in serial to avoid duplicate work
+    bool discovery_serializable = 9;
```
Comment (kumare3) [inline protos/flyteidl/core/tasks.proto:97]: i hate this name. I guess we want to keep it so.
but, ideally it should be cache_serializable right?

**Review by kumare3 — DISMISSED** [review]
one comment

File: protos/flyteidl/core/tasks.proto (hunk 2 above)
Comment (hamersaw) [reply protos/flyteidl/core/tasks.proto]: done. i'm ok with the switch, it hurt me to try and keep it uniform.

**Review by EngHabu — APPROVED** [review]

**Review by kumare3 — APPROVED** [review]
